go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -70,16 +74,72 @@ type serviceConfig struct {
+	var nodeportFlowsResult []string
+	if ipAddress != "" && nwDst != "" && nwSrc != "" {
+		// LB/EIP
+		nodeportFlowsResult = append(nodeportFlowsResult,
+			fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, %s=%s, tp_dst=%d, actions=ct(commit,zone=%d,table=6)",
+				cookie, npw.ofportPhys, flowProtocol, nwDst, ipAddress, svcPort, HostNodePortCTZone),
+			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, %s=%s, tp_src=%d, actions=ct(zone=%d,table=7)",
+				cookie, flowProtocol, nwSrc, ipAddress, svcPort, HostNodePortCTZone))
+	} else {
+		// NP
+		nodeportFlowsResult = append(nodeportFlowsResult,
+			// Traffic destined for ETP=local backed by OVN-K pod in LGW mode is matched on the svcNP and sent to table6.
+			fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=ct(commit,zone=%d,table=6)",
+				cookie, npw.ofportPhys, flowProtocol, svcPort, HostNodePortCTZone),
+			// Return traffic is matched on the svcNP as the sourcePort and gets un-DNATed back to nodeIP.
+			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, tp_src=%d, actions=ct(zone=%d,table=7)",
+				cookie, flowProtocol, svcPort, HostNodePortCTZone))
+	}
+	nodeportFlowsResult = append(nodeportFlowsResult,
+		// table 6, Sends the packet to Host. Note that the constant etp svc cookie is used since this flow would be
+		// same for all such services.
+		fmt.Sprintf("cookie=%s, priority=110, table=6, actions=output:LOCAL",
+			etpSvcOpenFlowCookie),
+		// table 7, Sends the reply packet back out eth0 to the external client. Note that the constant etp svc
+		// cookie is used since this would be same for all such services.
+		fmt.Sprintf("cookie=%s, priority=110, table=7, actions=output:%s",
+			etpSvcOpenFlowCookie, npw.ofportPhys))
+	return nodeportFlowsResult
 }
 
 // updateServiceFlowCache handles managing shared gateway flows for ingress traffic towards kubernetes services
 // (nodeport, external, ingress). By default incoming traffic into the node is steered directly into OVN.
 // If a service has externalTrafficPolicy local, and has host-networked endpoints, traffic instead will be steered directly
-// into the host.
+// into the host. If a service has externalTrafficPolicy local, and has non-host-networked endpoints and if its LGW mode,
+// traffic will be steered into LOCAL(to send it into ovn-k8s-mp0) preserving sourceIP.
 // add parameter indicates if the flows should exist or be removed from the cache
 // epHostLocal indicates if a host networked endpoint exists for this
+// etpLocalNonHostRules indicates if etp=local + hn=true + gw=local
 // service func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bool, epHostLocal bool) {
-func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bool, epHostLocal bool) {
+func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bool, epHostLocal, etpLocalNonHostRules bool) {
```

Comment:
again i think this should be separate parameters: epHostLocal, gw mode, etp=local